### PR TITLE
[Indexer] Create current token tables

### PR DESCRIPTION
**Pull request by bowenyang007**
### Description
* Cleaned up code a bit with types
* Added current token tables
* Make sure that for current token tables, only latest will be overridden (i.e. current transaction processing > last transaction stored in db)

### Test Plan
Ran locally and against devnet

Specifically to make sure that only latest will be overridden, I did extensive manual testing for now. 
1. Version 13760 has token X, amount 1
2. Ran indexer w/ mocked data: hardcoded to version 13761 with amount 1000, pulled token X and observed that it was updated
3. hardcoded to version 13761 with amount 9000, pulled token X and observed that it was *not* updated
4. hardcoded to version 13760 with amount 9000, pulled token X and observed that it was *not* updated
5. hardcoded to version 13770 with amount 9000, pulled token X and observed that it was updated

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4379)

### CODE REVIEW

File: crates/indexer/src/database.rs [hunk 1]
```diff
@@ -5,15 +5,23 @@
...
 
 pub type PgPool = diesel::r2d2::Pool<ConnectionManager<PgConnection>>;
 pub type PgDbPool = Arc<PgPool>;
 pub type PgPoolConnection = PooledConnection<ConnectionManager<PgConnection>>;
+#[derive(QueryId)]
+/// Using this will append a where clause at the end of the string upsert function, e.g.
+/// INSERT INTO ... ON CONFLICT DO UPDATE SET ... WHERE "transaction_version" = excluded."transaction_version"
+pub struct UpsertFilterLatestTransactionQuery<T> {
+    query: T,
+    where_clause: &'static str,
```
Comment (gregnazario) [inline crates/indexer/src/database.rs:23]: Interesting you aren't using an option

File: crates/indexer/src/database.rs [hunk 2]
```diff
@@ -60,19 +68,53 @@ pub fn execute_with_better_error<
...
+            query,
+            where_clause,
+        },
+        None => UpsertFilterLatestTransactionQuery {
+            query,
+            where_clause: "",
+        },
+    };
+    let debug = diesel::debug_query::<diesel::pg::Pg, _>(&final_query).to_string();
     aptos_logger::debug!("Executing query: {:?}", debug);
```
Comment (gregnazario) [inline crates/indexer/src/database.rs:93]: I guess this only prints out when debug is enabled?

You can also put this directly in the `aptos_logger::debug!`, and it's supposed to not even process it if the log is disabled.

File: crates/indexer/src/models/collection_datas.rs [hunk 3]
```diff
@@ -36,95 +39,90 @@ pub struct CollectionData {
...
+                                "version {} failed! collection creator resource was missing, table handle {} not in map {:?}",
+                                txn_version, TableMetadataForToken::standardize_handle(&table_handle), table_handle_to_owner,
+                            ))?;
+            let collection_data_id = CollectionDataIdType {
+                creator: creator_address,
+                name: collection_data.name,
+            };
+            let collection_data_id_hash = hash_str(&collection_data_id.to_string());
+            let collection_name = truncate_str(&collection_data_id.name, 128).to_string();
+            let metadata_uri = truncate_str(&collection_data.uri, 128).to_string();
```
Comment (gregnazario) [inline crates/indexer/src/models/collection_datas.rs:92]: 128 characters.  Isn't that a little short for a URI?

File: crates/indexer/src/models/tokens.rs [hunk 4]
```diff
@@ -145,20 +170,56 @@ impl Token {
...
-        (vec![], vec![], vec![], vec![])
+        (
+            vec![],
+            vec![],
+            vec![],
+            vec![],
+            HashMap::default(),
+            HashMap::default(),
+            HashMap::default(),
+        )
```
Comment (gregnazario) [inline crates/indexer/src/models/tokens.rs:222]: Btw, not sure, but I think you can get by with `Default::default()` and then get rid of the large tuple

**Review by gregnazario — COMMENTED** [review]
It looks sane to me.  But let's please add more comments & documentation after this PR.  It's a lot of functions that require a lot of context

**Review by gregnazario — APPROVED** [review]

File: crates/indexer/src/database.rs (hunk 1 above)
Comment (bowenyang007) [reply crates/indexer/src/database.rs:23]: Yea fair point I guess that we could.

File: crates/indexer/src/database.rs (hunk 2 above)
Comment (bowenyang007) [reply crates/indexer/src/database.rs:93]: good point.

File: crates/indexer/src/models/collection_datas.rs (hunk 3 above)
Comment (bowenyang007) [reply crates/indexer/src/models/collection_datas.rs:92]: Oh it should be 512. Good catch!

File: crates/indexer/src/processors/token_processor.rs [hunk 5]
```diff
@@ -195,8 +225,129 @@ fn insert_collection_datas(
...
+                    inserted_at.eq(excluded(inserted_at)),
+                )),
+            Some(" WHERE current_token_datas.last_transaction_version < excluded.last_transaction_version "),
+        ) {
+            Ok(_) => {}
+            Err(e) => {
+                return Err(e);
+            }
+        }
+    }
```
Comment (CapCap) [inline crates/indexer/src/processors/token_processor.rs:318]: `?`

File: crates/indexer/src/processors/token_processor.rs [hunk 6]
```diff
@@ -223,15 +374,59 @@ impl TransactionProcessor for TokenTransactionProcessor {
         let mut all_token_ownerships = vec![];
         let mut all_token_datas = vec![];
         let mut all_collection_datas = vec![];
+
+        let mut all_current_token_ownerships: HashMap<
+            CurrentTokenOwnershipPK,
+            CurrentTokenOwnership,
+        > = HashMap::new();
+        let mut all_current_token_datas: HashMap<TokenDataIdHash, CurrentTokenData> =
```
Comment (CapCap) [inline crates/indexer/src/processors/token_processor.rs:382]: add comments to this pls

**Review by CapCap — APPROVED** [review]

**Pull request merged**